go/vt/topo/stats_conn.go

```diff
@@ -46,21 +53,29 @@ type StatsConn struct {
 	cell     string
 	conn     Conn
 	readOnly bool
+	readSem  *semaphore.Weighted
 }
 
 // NewStatsConn returns a StatsConn
-func NewStatsConn(cell string, conn Conn) *StatsConn {
+func NewStatsConn(cell string, conn Conn, readSem *semaphore.Weighted) *StatsConn {
 	return &StatsConn{
 		cell:     cell,
 		conn:     conn,
 		readOnly: false,
+		readSem:  readSem,
 	}
 }
 
 // ListDir is part of the Conn interface
 func (st *StatsConn) ListDir(ctx context.Context, dirPath string, full bool) ([]DirEntry, error) {
 	startTime := time.Now()
 	statsKey := []string{"ListDir", st.cell}
+	if err := st.readSem.Acquire(ctx, 1); err != nil {
+		return nil, err
+	}
+	defer st.readSem.Release(1)
+	topoStatsConnReadWaitTimings.Record(statsKey, startTime)
+	startTime = time.Now() // reset
```

Comment:
I think that we should remove this second setting of startTime and move the recording *before* we try to acquire the semaphore. The timer is meant to measure how long the ListDir call takes and IMO that's what it should do — which now includes acquiring the semaphore. Not only is that more accurate/correct, but it could help identify unexpected long ListDir times which could likely be improved by increasing the default read concurrency.

Same thing for the other functions below.

Or am I missing something here?

Reply:
@mattlord my goal here is be able to measure the "operation" and "semaphore wait" timing separately so it's easy to break down if operations are waiting on the local semaphore or the server

With the current approach `topoStatsConnReadWaitTimings` will tell us if the default read concurrency is too low and due to resetting `startTime` the `topoStatsConnTimings` tells us purely the post-semaphore client<->server timing

So I suppose I intentionally didn't include the semaphore wait time in `topoStatsConnTimings` because it has it's own metric, but I see your point. The consequence of including the semaphore wait time in `topoStatsConnTimings` is you would need to subtract `topoStatsConnReadWaitTimings` to know if the server or network is responsible for slowness

Reply:
@timvaillancourt You're not measuring anything separately though. 

There's only one stat updated with a value representing the recorded duration of the call, and that is `ListDir-<cell>`. The startTime is passed in by value, so later updated the startTime has no effect.

I also don't see why you're doing this: 
```
topoStatsConnReadWaitTimings = stats.NewMultiTimings(
		"TopologyConnReadWaits",
		"TopologyConnReadWait timings",
		[]string{"Operation", "Cell"})
```

Since the cell dimension is already captured in the current key. I don't see how this adds any dimensions or value, AND this could alter the relevant stats at the `/metrics` endpoint which we probably don't want to do as it would potentially mess up existing monitoring (grafana dashboards e.g.) code that uses the old metric names.

One of us is mistaken / confused here, and it could well be me. 😄 Could you please share some of the resulting stats from `/metrics` and/or `/debug/stats` for these? And assuming that it's all working as intended, could you please add some comments to the code to make the reasoning/intention clear?

Thanks!

Reply:
One other related note, IF we are indeed capturing two things for each call — the topo call time and the semaphore wait time — then we should have a unit test that confirms this works as expected.

Reply:
OK, coming back to this... I can see what you were trying to do. You left the defer in as it was, so that key and its behavior/timing remains unchanged. Sorry, I was confused by the original diff output in the GH UI.

So we have two stats created for each call now. The original one, and a new one:
```
topoStatsConnReadWaitTimings = stats.NewMultiTimings(
		"TopologyConnReadWaits",
		"TopologyConnReadWait timings",
		[]string{"Operation", "Cell"})
```

And rather than recording the ListDir time, in this case, it's recording the read wait time. With that wait being how long it takes for us to acquire the semaphore.

That makes sense! I'm sorry for being so dense previously. 🙂 I should have just checked out the PR and it would have been obvious from the get go. That's my fault.

I still think it's probably worth adding something to an existing unit test to confirm the new stats, but it looks good to me! Thanks again!